src/v/cluster/topics_frontend.cc

```diff
@@ -794,4 +795,97 @@ ss::future<topic_result> topics_frontend::do_create_partition(
+
+    // no leader available
+    if (!leader) {
+        co_return errc::no_leader_controller;
+    }
+    if (leader == _self) {
+        if (dir == partition_move_direction::from_node) {
+            co_return co_await do_cancel_moving_partition_replicas(
+              _topics.local().ntps_moving_from_node(node_id), timeout);
+        }
+
+        co_return co_await do_cancel_moving_partition_replicas(
+          _topics.local().ntps_moving_to_node(node_id), timeout);
+    }
+
+    co_return co_await _connections.local()
+      .with_node_client<controller_client_protocol>(
+        _self,
+        ss::this_shard_id(),
+        *leader,
+        timeout,
+        [timeout, node_id, dir](controller_client_protocol client) mutable {
+            return client
+              .cancel_node_partition_movements(
+                cancel_node_partition_movements_request{
+                  .node_id = node_id, .direction = dir},
+                rpc::client_opts(timeout))
+              .then(&rpc::get_ctx_data<cancel_partition_movements_reply>);
+        })
+      .then([](result<cancel_partition_movements_reply> r) {
+          return r.has_error() ? ret_t(r.error())
+                               : std::move(r.value().partition_results);
+      });
+}
+
+ss::future<result<std::vector<partition_result>>>
+topics_frontend::cancel_moving_all_partition_replicas(
+  model::timeout_clock::time_point timeout) {
+    using ret_t = result<std::vector<partition_result>>;
+    auto leader = _leaders.local().get_leader(model::controller_ntp);
+
+    // no leader available
+    if (!leader) {
+        co_return errc::no_leader_controller;
+    }
+    if (leader == _self) {
+        co_return co_await do_cancel_moving_partition_replicas(
+          _topics.local().all_updates_in_progress(), timeout);
+    }
+
+    co_return co_await _connections.local()
+      .with_node_client<controller_client_protocol>(
+        _self,
+        ss::this_shard_id(),
+        *leader,
+        timeout,
+        [timeout](controller_client_protocol client) mutable {
+            return client
+              .cancel_all_partition_movements(
+                cancel_all_partition_movements_request{},
+                rpc::client_opts(timeout))
+              .then(&rpc::get_ctx_data<cancel_partition_movements_reply>);
+        })
+      .then([](result<cancel_partition_movements_reply> r) {
+          return r.has_error() ? ret_t(r.error())
+                               : std::move(r.value().partition_results);
+      });
+}
+
+ss::future<std::vector<partition_result>>
+topics_frontend::do_cancel_moving_partition_replicas(
+  std::vector<model::ntp> ntps, model::timeout_clock::time_point timeout) {
+    std::vector<partition_result> results;
+    results.reserve(ntps.size());
+    co_await ss::max_concurrent_for_each(
+      ntps, 32, [this, &results, timeout](model::ntp& ntp) {
+          return cancel_moving_partition_replicas(ntp, timeout)
```

Comment:
So how are we going to handle decommissioning nodes? I guess one option is to do nothing - the user can always manually recommission later to use the node again, otherwise it will just remain in the decommissioning state.